Python/ceval.c

```diff
@@ -2702,6 +2702,10 @@ void Py_LeaveRecursiveCall(void)
 
 ///////////////////// Experimental UOp Interpreter /////////////////////
 
+#undef JUMP_POP_DISPATCH
+#define JUMP_POP_DISPATCH(x, n) \
```

Comment:
This seems unnecessary.

The superblock generator should be converting each conditional jump into a conditional exit, followed by an unconditional jump.
Such that:
```
POP_JUMP_IF_TRUE label
```
becomes:

#### False branch more likely
```
EXIT_IF_TRUE
```

#### True branch more likely
```
EXIT_IF_FALSE
JUMP_FORWARD label
```

As for which branch is more likely, we will need to record which way branches go.

Reply:
Okay, just so I understand correctly, the idea is that `EXIT_IF_TRUE`/`FALSE` doesn't pop, and continues in the bytecode at the original `POP_JUMP_IF_TRUE` instruction, which will pop. So the "False branch more likely" (i.e., branch not likely) version will have to actually translate to `EXIT_IF_TRUE; POP_TOP`. We could special-case this in the translator: it would have put such a list of uops in the expansion for `POP_JUMP_IF_TRUE` and friends, and we'd have to add hand-written cases to the uop executor for `EXIT_IF_TRUE` etc.

In the other case (branch likely) the `JUMP_FORWARD label` could just be `SAVE_IP label`; superblock generation would then continue from the bytecode at `label`. That's not something I currently do -- a simplistic approach could do the `SAVE_IP` followed by `EXIT_TRACE`, and we can iterate later on continuing from `label`.

But recording which way branches go is something for a future PR; again the most simplistic thing to do for now is to assume that branching is less likely than not branching (i.e., always generate `EXIT_IF_TRUE; POP_TOP` for now).

Honestly, for now I think I'll stick to just making the generator explicitly translate `JUMPBY(n)` into the correct sequence of `JUMPBY`, `STACK_SHRINK` and `DISPATCH`. Nothing should follow `JUMPBY` then.

Reply:
I think it makes more sense to consume the condition before exiting

These two instructions, `POP_JUMP_IF_TRUE` and `POP_JUMP_IF_FALSE` (the `None` variants can be broken down into `LOAD_CONST None; IS_OP; POP_JUMP...`) are special, so don't worry too much about fitting them into the more general framework.

There are a number of approaches, for exits. Here are three:

1. Leave the condition on the stack, and exit to the `POP_JUMP...` instruction.
    I don't like this approach as it is inefficient and will make trace stitching tricky
2. Pop the condition and exit to the target, or successor, instruction.
    This is better, but means that the `EXIT` uop needs to handle a jump and the exit.
3. Convert the superblock into a tree, jumping to an exit branch which makes an unconditional exit.
    Definitely my preferred option. Means that there is only one `EXIT` uop and that it just exits. 

Option 3 makes the superblock creation more complex, but simplifies the tier2 interpreter.
We will want to push some operations onto exit branches in the specializer and partial evaluator, so we will need to convert the superblock to a tree at some point.

If option 3 is too complex for this PR, option 2 should work as a temporary step.

So for option 3, using the example above of `POP_JUMP_IF_TRUE label` we get:

## False branch more likely
```
    POP_JUMP_IF_TRUE_UOP uop_label:
    ...
uop_label:
   SAVE_IP label
   EXIT
```

#### True branch more likely
```
   POP_JUMP_IF_FALSE_UOP uop_label:
   SAVE_IP label
uop_label:
    EXIT
```